File: android/services/yandex/src/main/java/com/mplayeraudio/services/yandex/YandexMusicRepositoryImpl.kt
package com.mplayeraudio.services.yandex

import com.mplayeraudio.core.domain.musiclibrary.MusicLibraryRepository
import com.mplayeraudio.core.domain.musiclibrary.MusicServiceAvailability
import com.mplayeraudio.core.domain.musiclibrary.Playlist
import com.mplayeraudio.core.domain.musiclibrary.PlaylistId
import com.mplayeraudio.core.domain.musiclibrary.PlaylistSummary
import com.mplayeraudio.core.domain.musiclibrary.SavedTracksResult
import com.mplayeraudio.core.domain.musiclibrary.Track
import com.mplayeraudio.core.domain.musiclibrary.TrackRef
import com.mplayeraudio.services.yandex.internal.YandexMusicRequestRunner
import com.mplayeraudio.services.yandex.internal.toApiTrackId
import com.mplayeraudio.services.yandex.internal.toAvailability
import com.mplayeraudio.services.yandex.internal.toCurrentUserId
import com.mplayeraudio.services.yandex.internal.toPlaylist
import com.mplayeraudio.services.yandex.internal.toPlaylistSummary
import com.mplayeraudio.services.yandex.internal.toSavedTracksResult
import com.mplayeraudio.services.yandex.internal.toTrack
import com.mplayeraudio.services.yandex.internal.network.YandexMusicApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.jsonObject

internal class YandexMusicRepositoryImpl(
    private val requestRunner: YandexMusicRequestRunner,
    private val api: YandexMusicApi,
) : MusicLibraryRepository {
    private val availabilityState = MutableStateFlow<MusicServiceAvailability?>(null)
    private val ownPlaylistsState = MutableStateFlow<List<PlaylistSummary>>(emptyList())
    private val playlistState = MutableStateFlow<Map<PlaylistId, Playlist>>(emptyMap())
    private val savedTracksState = MutableStateFlow<SavedTracksResult?>(null)
    private val trackState = MutableStateFlow<Map<TrackRef, Track>>(emptyMap())

    override fun observeAvailability(): Flow<MusicServiceAvailability> {
        return availabilityState.filterNotNull()
    }

    override fun observeOwnPlaylists(): Flow<List<PlaylistSummary>> {
        return ownPlaylistsState
    }

    override fun observePlaylist(id: PlaylistId): Flow<Playlist?> {
        return playlistState.map { playlists -> playlists[id] }
    }

    override fun observeSavedTracks(): Flow<SavedTracksResult> {
        return savedTracksState.filterNotNull()
    }

    override fun observeTracks(refs: List<TrackRef>): Flow<List<Track>> {
        return trackState.map { tracks ->
            refs.mapNotNull(tracks::get)
        }
    }

    override suspend fun refreshAvailability() {
        availabilityState.value = requestRunner.withAuthorizedRequest { accessToken ->
            api.fetchAvailability(accessToken).toAvailability()
        }
    }

    override suspend fun refreshOwnPlaylists() {
        ownPlaylistsState.value = requestRunner.withCurrentUserId { accessToken, userId ->
            api.fetchOwnPlaylists(
                accessToken = accessToken,
                userId = userId.value,
            ).map { element -> element.jsonObject.toPlaylistSummary() }
        }
    }

    override suspend fun refreshPlaylist(id: PlaylistId) {
        val playlist = requestRunner.withAuthorizedRequest { accessToken ->
            api.fetchPlaylist(
                accessToken = accessToken,
                userId = id.ownerId.value,
                kind = id.kind.value,
            ).toPlaylist()
        }
        playlistState.value = playlistState.value + (id to playlist)
        cacheTracks(playlist.tracks.mapNotNull { entry -> entry.track })
    }

    override suspend fun refreshSavedTracks() {
        savedTracksState.value = requestRunner.withCurrentUserId { accessToken, userId ->
            api.fetchSavedTracks(
                accessToken = accessToken,
                userId = userId.value,
            ).toSavedTracksResult()
        }
    }

    override suspend fun refreshTracks(refs: List<TrackRef>) {
        if (refs.isEmpty()) {
            return
        }

        val loadedTracks = requestRunner.withAuthorizedRequest { accessToken ->
            api.fetchTracks(
                accessToken = accessToken,
                trackIds = refs.map(TrackRef::toApiTrackId),
            ).map { element -> element.jsonObject.toTrack() }
        }

        val updatedTracks = trackState.value.toMutableMap()
        loadedTracks.forEach { track ->
            updatedTracks[track.preview.ref] = track
        }
        trackState.value = updatedTracks
    }

    private fun cacheTracks(tracks: List<Track>) {
        if (tracks.isEmpty()) {
            return
        }

        val updatedTracks = trackState.value.toMutableMap()
        tracks.forEach { track ->
            updatedTracks[track.preview.ref] = track
        }
        trackState.value = updatedTracks
    }
}

File: ios/Services/YandexMusic/Sources/YandexMusicService/YandexMusicRepositoryImpl.swift
import AuthFeature
import CoreDomain
import Foundation

// MARK: - Repository

public final class YandexMusicProvider: MusicProvider, @unchecked Sendable {
    public let id: MusicProviderId = .yandexMusic

    private let requestRunner: YandexMusicRequestRunner

    private let availabilityState = AsyncValueRelay<MusicServiceAvailability?>(nil)
    private let playlistsState = AsyncValueRelay<[PlaylistSummary]>([])
    private let playlistState = AsyncValueRelay<[PlaylistId: Playlist]>([:])
    private let savedTracksState = AsyncValueRelay<SavedTracksResult?>(nil)
    private let trackState = AsyncValueRelay<[TrackRef: Track]>([:])

    public init(
        accessTokenProvider: YandexAccessTokenProvider,
        api: YandexMusicAPI,
        urlSession: URLSession = .shared
    ) {
        requestRunner = YandexMusicRequestRunner(
            accessTokenProvider: accessTokenProvider,
            api: api
        )
    }

    public convenience init(
        accessTokenProvider: YandexAccessTokenProvider,
        urlSession: URLSession = .shared
    ) {
        self.init(
            accessTokenProvider: accessTokenProvider,
            api: URLSessionYandexMusicAPI(session: urlSession)
        )
    }

    // MARK: - Observe

    public func observeAvailability() -> AsyncStream<MusicServiceAvailability> {
        filterNonNil(availabilityState)
    }

    public func observePlaylists() -> AsyncStream<[PlaylistSummary]> {
        playlistsState.stream()
    }

    public func observePlaylist(id: PlaylistId) -> AsyncStream<Playlist?> {
        mapped(playlistState) { $0[id] }
    }

    public func observeSavedTracks() -> AsyncStream<SavedTracksResult> {
        filterNonNil(savedTracksState)
    }

    public func observeTracks(refs: [TrackRef]) -> AsyncStream<[Track]> {
        mapped(trackState) { tracks in refs.compactMap { tracks[$0] } }
    }

    // MARK: - Refresh

    public func refreshAvailability() async throws {
        let availability = try await requestRunner.withAuthorizedRequest { accessToken in
            let json = try await self.requestRunner.api.fetchAvailability(accessToken: accessToken)
            return try json.toAvailability()
        }
        availabilityState.yield(availability)
    }

    public func refreshPlaylists() async throws {
        let summaries = try await requestRunner.withCurrentUserId { accessToken, userId in
            let items = try await self.requestRunner.api.fetchOwnPlaylists(
                accessToken: accessToken,
                userId: userId.rawValue
            )
            return try items.map { try $0.toPlaylistSummary() }
        }
        playlistsState.yield(summaries)
    }

    public func refreshPlaylist(id: PlaylistId) async throws {
        let playlist = try await requestRunner.withAuthorizedRequest { accessToken in
            let json = try await self.requestRunner.api.fetchPlaylist(
                accessToken: accessToken,
                userId: id.ownerId.rawValue,
                kind: id.kind.rawValue
            )
            return try json.toPlaylist()
        }
        playlistState.yield(playlistState.currentValue.merging([id: playlist]) { $1 })
        cacheTracks(playlist.tracks.compactMap(\.track))
    }

    public func refreshSavedTracks() async throws {
        let result = try await requestRunner.withCurrentUserId { accessToken, userId in
            let raw = try await self.requestRunner.api.fetchSavedTracks(
                accessToken: accessToken,
                userId: userId.rawValue
            )
            return try parseSavedTracksResult(from: raw, userId: userId)
        }
        savedTracksState.yield(result)
    }

    public func refreshTracks(refs: [TrackRef]) async throws {
        guard !refs.isEmpty else { return }
        let loaded = try await requestRunner.withAuthorizedRequest { accessToken in
            let items = try await self.requestRunner.api.fetchTracks(
                accessToken: accessToken,
                trackIds: refs.map { $0.toApiTrackId() }
            )
            return try items.map { try $0.toTrack() }
        }
        cacheTracks(loaded)
    }

    // MARK: - Private helpers

    private func cacheTracks(_ tracks: [Track]) {
        guard !tracks.isEmpty else { return }
        let additions = Dictionary(
            tracks.map { ($0.preview.ref, $0) },
            uniquingKeysWith: { $1 }
        )
        trackState.yield(trackState.currentValue.merging(additions) { $1 })
    }

    private func filterNonNil<T: Sendable>(
        _ relay: AsyncValueRelay<T?>
    ) -> AsyncStream<T> {
        AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            let task = Task {
                for await value in relay.stream() {
                    guard let value else { continue }
                    continuation.yield(value)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func mapped<Source: Sendable, Element: Sendable>(
        _ relay: AsyncValueRelay<Source>,
        transform: @escaping @Sendable (Source) -> Element
    ) -> AsyncStream<Element> {
        AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            let task = Task {
                for await value in relay.stream() {
                    continuation.yield(transform(value))
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

// MARK: - Request runner

final class YandexMusicRequestRunner: @unchecked Sendable {
    let api: YandexMusicAPI
    private let accessTokenProvider: YandexAccessTokenProvider
    private let lock = NSLock()
    private var _cachedUserId: ProviderUserId?

    init(accessTokenProvider: YandexAccessTokenProvider, api: YandexMusicAPI) {
        self.accessTokenProvider = accessTokenProvider
        self.api = api
    }

    func withCurrentUserId<T: Sendable>(
        _ block: @Sendable (String, ProviderUserId) async throws -> T
    ) async throws -> T {
        try await withAuthorizedRequest { accessToken in
            let userId = try await self.requireCurrentUserId(accessToken: accessToken)
            return try await block(accessToken, userId)
        }
    }

    func withAuthorizedRequest<T: Sendable>(
        _ block: @Sendable (String) async throws -> T
    ) async throws -> T {
        let token = try await accessTokenProvider.validAccessToken(forceRefresh: false)
        do {
            return try await block(token)
        } catch MusicLibraryError.unauthorized {
            let freshToken = try await accessTokenProvider.validAccessToken(forceRefresh: true)
            return try await block(freshToken)
        }
    }

    private func requireCurrentUserId(accessToken: String) async throws -> ProviderUserId {
        if let cached = lock.withLock({ _cachedUserId }) {
            return cached
        }
        let json = try await api.fetchCurrentUser(accessToken: accessToken)
        let userId = try json.toCurrentUserId()
        lock.withLock { _cachedUserId = userId }
        return userId
    }
}
